File: 450-DSA-SHEET/ARRAYS/C-Plus-Plus/5-Move_all_negative_to_one_side.cpp
#include <bits/stdc++.h>
using namespace std;

void swap(int &a, int &b){
    int temp = a;
    a = b;
    b = temp;
}

void move(int arr[], int n){
    // USING DUTCH NATIONAL FLAG CONCEPT
    int low = 0;
    int high = n-1;
    while(low < high){
        if(arr[low] < 0){
            low++;
        }
        else if(arr[high] > 0){
            high--;
        }
        else{
            swap(arr[low],arr[high]);
        }
    }
}


//-------------------------------------------------------------

// APPROACH 1

/*
int main(){
    int n; 
    cout << "Enter the number of elements: ";
    cin>>n;
    int arr[n];
    cout << "Enter the elements of the array: ";
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    sort(arr,arr+n);
    cout << "Moving all the elements to one side: " << endl;
    for(int i=0;i<n;i++){
        cout << arr[i] << " ";
    }
    cout << endl;
    return 0;
}
*/

//-------------------------------------------------------------

// APPROACH 2: BY USING DNF SORT CONCEPT MODIFICATION OF IT

int main(){
    int n; 
    cout << "Enter the number of elements: ";
    cin>>n;
    int arr[n];
    cout << "Enter the elements of the array: ";
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    move(arr,n);
    cout << "Moving all the elements to one side: " << endl;
    for(int i=0;i<n;i++){
        cout << arr[i] << " ";
    }
    cout << endl;
    return 0;
}

File: 450-DSA-SHEET/ARRAYS/Python/5-Move_all_negative_to_one_side.py

# APPROACH 1

"""
def move(arr):
    arr.sort()

arr = [-1, 2, -3, 4, 5, 6, -7, 8, 9]
move(arr)
for i in arr:
    print(i, end = ' ')

"""

# ----------------------------------------------------

# APPROACH 2

def move(arr, n):
    low = 0
    high = n-1
    while(low<high):
        if (arr[low]<0):
            low+=1
        elif (arr[high]>0):
            high-=1
        else:
            arr[low],arr[high] = arr[high],arr[low]


def printArray(arr, n):
    for i in range(n):
        print(arr[i], end = ' ')

arr = [1, 2,  -4, -5, 2, -7, 3, 2, -6, -8, -9, 3, 2, 1]
n = len(arr)
move(arr, n)
printArray(arr, n)